### PR TITLE
Add `BillboardTeaserBlock`

### PR DESCRIPTION
**Pull request by SebiVPS**
---

## PR Checklist

-   [x] Link to the respective task if one exists: COM-838
-   [x] Provide screenshots/screencasts if the change contains visual changes
-   [x] Add overlay

<details>
    <summary>Screenshots/screencasts</summary>
Screenshot 2024-07-18 at 16 18 34

</details>

### CODE REVIEW

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    flex: 1;
+    flex-direction: column;
+    justify-content: center;
+    align-items: ${({ $alignItems }) => $alignItems};
+`;
+
+const BaseImage = styled.div`
+    img {
+        object-fit: cover;
+    }
```
Comment (jamesricky) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:76]: This should probably be handled by the Image-/Media-Block itself.

Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:76]: yeah, need to check that with the updated image/media component

**Comment by nsams** [comment]
shouldn't the content be aligned just like other content? (becasue you don't use the page grid)

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    ),
+    { label: "Billboard Teaser" },
+);
+
+const Root = styled.div`
+    position: relative;
+    overflow: hidden;
+`;
+
+const AbsoluteGridRoot = styled(PageLayout)`
```
Comment (SebiVPS) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:48]: @nsams it is used here

**Review by jamesricky — APPROVED** [review]

**Comment by johnnyomair** [comment]
Please change the base to `main`.

**Review by jamesricky — DISMISSED** [review]

File: admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        displayName: <FormattedMessage id="billboardTeaserBlock.displayName" defaultMessage="Billboard Teaser" />,
+        blocks: {
+            media: {
+                block: MediaBlock,
+                title: <FormattedMessage id="billboardTeaserBlock.media" defaultMessage="Media" />,
+                hiddenInSubroute: true,
+            },
+            heading: {
+                block: HeadingBlock,
+            },
```
Comment (johnnyomair) [inline admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx:21]: This probably needs a hiddenInSubroute as well

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                <MediaBlock data={media} aspectRatio="16x9" />
+            </ImageDesktop>
+            <ImageLargeDesktop>
+                <MediaBlock data={media} aspectRatio="3x1" />
+            </ImageLargeDesktop>
+            <AbsoluteGridRoot grid>
+                <Content $alignItems={alignmentMap[heading.textAlignment]}>
+                    <HeadingBlock data={heading} />
+                    <RichTextBlock data={text} />
+                    {callToActionList.blocks.length > 0 && <CallToActionListBlock data={callToActionList} />}
```
Comment (johnnyomair) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:35]: Shouldn't this length check be inside CallToActionListBlock?

File: admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 3 above)
Comment (SebiVPS) [reply admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx:21]: I do not know exactly why, but it does not need ahiddenInSubroute

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 4 above)
Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx]: removed it

File: site/src/common/blocks/RichTextBlock.tsx [hunk 5]
```diff
@@ -1,94 +1,60 @@
...
+    SUB: (children, { key }) => <sub key={key}>{children}</sub>,
+    SUP: (children, { key }) => <sup key={key}>{children}</sup>,
+    STRIKETHROUGH: (children, { key }) => <s key={key}>{children}</s>,
+};
+
 /**
  * Define the renderers
  */
-export const defaultRichTextRenderers: Renderers = {
+const getDefaultRenderers = (colorInverted?: boolean): Renderers => ({
```
Comment (johnnyomair) [inline site/src/common/blocks/RichTextBlock.tsx:33]: In #282 @jomunker removed the default text color from the rich text block, which is a good idea. Then the inverted color could be set in the billboard teaser block.

File: site/src/common/components/Typography.tsx [hunk 6]
```diff
@@ -195,23 +195,31 @@ export interface TypographyProps extends HTMLAttributes<HTMLElement> {
...
+        {...restProps}
+    >
         {children}
     </Text>
 );
 
 interface TextProps {
     $variant: TypographyVariant;
     $bottomSpacing?: boolean;
+    $colorInverted?: boolean;
```
Comment (johnnyomair) [inline site/src/common/components/Typography.tsx:217]: Typography shouldn't set a color at all, then this could be done in the billboard teaser block

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    padding: ${({ theme }) => theme.spacing.D200} 0;
+    display: flex;
+    flex: 1;
+    flex-direction: column;
+    justify-content: center;
+    align-items: ${({ $alignItems }) => $alignItems};
+`;
+
+const BaseImage = styled.div<{ $backgroundOpacity: string }>`
+    filter: brightness(calc(100% - ${({ $backgroundOpacity }) => $backgroundOpacity}%));
```
Comment (johnnyomair) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:67]: You could also set a transparent background on the AbsoluteGridRoot component. I don't know what's better though 🤷🏼‍♂️

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+import { MediaBlock } from "@src/common/blocks/MediaBlock";
+import { RichTextBlock } from "@src/common/blocks/RichTextBlock";
+import { PageLayout } from "@src/layout/PageLayout";
+import { CSSProperties } from "react";
+import styled from "styled-components";
+
+const alignmentMap: Record<HeadingBlockData["textAlignment"], CSSProperties["alignItems"]> = {
+    Left: "left",
+    Center: "center",
+};
```
Comment (johnnyomair) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:14]: Using the headline alignment to align everything seems arbitrary to me... Did you discuss this with UX?

File: api/src/documents/pages/blocks/billboard-teaser.block.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    percentage90 = "90",
+    percentage80 = "80",
+    percentage70 = "70",
+    percentage60 = "60",
+    percentage50 = "50",
+    percentage40 = "40",
+    percentage30 = "30",
+    percentage20 = "20",
+    percentage10 = "10",
+}
```
Comment (johnnyomair) [inline api/src/documents/pages/blocks/billboard-teaser.block.ts:29]: Instead of an enum you could also use number or string 🤔

**Review by mariokemetinger — APPROVED** [review]

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 8 above)
Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:14]: For me it seems superfluous to have a second  alignment select with same options. But if thats what we want i'll add that.

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 7 above)
Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:68]: I do not know what's better either ... @jamesricky @mariokemetinger ?

File: api/src/documents/pages/blocks/billboard-teaser.block.ts (hunk 9 above)
Comment (SebiVPS) [reply api/src/documents/pages/blocks/billboard-teaser.block.ts]: i think number is not supported by select field in admin, ill make it a string then 👍

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                </Content>
+            </AbsoluteGridRoot>
+        </Root>
+    ),
+    { label: "Billboard Teaser" },
+);
+
+const Root = styled(PageLayout)`
+    position: relative;
+    overflow: hidden;
```
Comment (jamesricky) [inline site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:40]: I don't think this should be needed 🤔

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 7 above)
Comment (jamesricky) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx:63]: I'd just add a pseudo-element that overlays the image with a black `background-color` what has a variable `opacity` value.

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 10 above)
Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx]: if someone adds a lot of content it will overflow, but thats content ...

**Comment by SebiVPS** [comment]
> I'd just add a pseudo-element that overlays the image with a black background-color what has a variable opacity value.

i found the one liner quite elegant

File: site/src/documents/pages/blocks/BillboardTeaserBlock.tsx (hunk 8 above)
Comment (SebiVPS) [reply site/src/documents/pages/blocks/BillboardTeaserBlock.tsx]: removed that, according to the design the content should always be centered

**Review by jamesricky — DISMISSED** [review]

File: site/src/common/blocks/HeadingBlock.tsx [hunk 11]
```diff
@@ -7,51 +7,21 @@ import { CSSProperties } from "react";
 import { Renderers } from "redraft";
 import styled from "styled-components";
 
-import { defaultRichTextRenderers, RichTextBlock } from "./RichTextBlock";
+import { createTextBlockRenderFn, defaultRichTextInlineStyleMap, RichTextBlock } from "./RichTextBlock";
 
 const eyebrowRenderers: Renderers = {
-    inline: defaultRichTextRenderers.inline,
+    inline: defaultRichTextInlineStyleMap,
```
Comment (johnnyomair) [inline site/src/common/blocks/HeadingBlock.tsx:13]: Why did you change this?

File: api/src/documents/pages/blocks/billboard-teaser.block.ts (hunk 9 above)
Comment (johnnyomair) [reply api/src/documents/pages/blocks/billboard-teaser.block.ts]: Number is supported, and better for validation: 2087247d8464318a4617782be3971f9593401d05

File: admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    { value: 100, label: "100%" },
+    { value: 90, label: "90%" },
+    { value: 80, label: "80%" },
+    { value: 70, label: "70%" },
+    { value: 60, label: "60%" },
+    { value: 50, label: "50%" },
+    { value: 40, label: "40%" },
+    { value: 30, label: "30%" },
+    { value: 20, label: "20%" },
+    { value: 10, label: "10%" },
```
Comment (thomasdax98) [inline admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx:20]: 0 - 90 instead of 10 - 100

Comment (thomasdax98) [reply admin/src/documents/pages/blocks/BillboardTeaserBlock.tsx]: 741189f6f99e8eb1ce0afdda4c651f6bda15c455

File: site/src/common/blocks/HeadingBlock.tsx (hunk 11 above)
Comment (thomasdax98) [reply site/src/common/blocks/HeadingBlock.tsx:13]: Unfortunately, I can't comment on that. So we have to wait for @SebiVPS

Comment (SebiVPS) [reply site/src/common/blocks/HeadingBlock.tsx:13]: This changes has been made during necessary code refactoring when adding the `colorInverted` prop. The prop was then removed again when the text color was added to the gobal styles.

These changes do not need to be part of this PR. I think the changes make the code a little bit more readable and usable. As they are indeed no part of this PR you can remove them and, if you want, make a new PR with these adaptions?

**Comment by johnnyomair** [comment]
I'm not sure if the background opacity makes sense from a UX perspective:

If I change it to 10%, the background is very visible:

background opacity 10

If I change it to 90%, the background is barely visible:

background opacity 90

Shouldn't this be the other way around?

**Comment by thomasdax98** [comment]
@SebiVPS Please rename "background opacity" to "overlay". 

"background opacity" is unintuitive because 0 means full opacity and 90 means almost no opacity.

**Review by jamesricky — APPROVED** [review]

File: site/src/common/blocks/HeadingBlock.tsx (hunk 11 above)
Comment (johnnyomair) [reply site/src/common/blocks/HeadingBlock.tsx:13]: We can keep it in this PR

**Review by johnnyomair — APPROVED** [review]

**Review by thomasdax98 — APPROVED** [review]

**Review by mariokemetinger — APPROVED** [review]
